Make sure that disabled jobs are paused in Quartz

machi1990 opened the pull request:
Fixes https://github.com/quarkusio/quarkus/issues/25214

Disabled methods whose jobs details and triggers were already in the
database were still getting triggered because Quartz invokes all the
jobs in the database that are not in the paused State.
This is to make sure that the disabled methods are effectively not
triggered by Quartz in any way.

Furthermore, I do not believe the full scheduled method implementing the Application custom logic
was called because of the checks in the extensions which meant the
App business logic trigger was not full invoked but pausing the triggers
via Quartz we are sure that the trigger won't be invoked when it is
disabled.

Previously this was done via checks in the extension which mean that
only the Application logic of the scheduled method was not called.

- https://github.com/quarkusio/quarkus/blob/4df39b60a0b3d132713e6d60087a1054588d4a5f/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L215-L218
- https://github.com/quarkusio/quarkus/blob/4df39b60a0b3d132713e6d60087a1054588d4a5f/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L168-L170

which meant that the trigger was not stored in known business triggers
at https://github.com/quarkusio/quarkus/blob/4df39b60a0b3d132713e6d60087a1054588d4a5f/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L302-L303

meaning that upon the execution invocation of the method in
https://github.com/quarkusio/quarkus/blob/4df39b60a0b3d132713e6d60087a1054588d4a5f/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L560
the business logic wasn't triggered but a NPE (same as https://github.com/quarkusio/quarkus/issues/24931) could have been raised
with the code in main branch

machi1990 committed:
fix(quartz): make sure that disabled jobs are paused

Fixes https://github.com/quarkusio/quarkus/issues/25214

Disabled methods whose jobs details and triggers were already in the
database were still getting triggered because Quartz invokes all the
jobs in the database that are not in the paused State.
This is to make sure that the disabled methods are effectively not
triggered by Quartz in any way.

Furthermore, I do not believe the full scheduled method implementing the Application custom logic
was called because of the checks in the extensions which meant the
App business logic trigger was not full invoked but pausing the triggers
via Quartz we are sure that the trigger won't be invoked when it is
disabled.

Previously this was done via checks in the extension which mean that
only the Application logic of the scheduled method was not called.

- https://github.com/quarkusio/quarkus/blob/4df39b60a0b3d132713e6d60087a1054588d4a5f/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L215-L218
- https://github.com/quarkusio/quarkus/blob/4df39b60a0b3d132713e6d60087a1054588d4a5f/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L168-L170

which meant that the trigger was not stored in known business triggers
at https://github.com/quarkusio/quarkus/blob/4df39b60a0b3d132713e6d60087a1054588d4a5f/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L302-L303

meaning that upon the execution invocation of the method in
https://github.com/quarkusio/quarkus/blob/4df39b60a0b3d132713e6d60087a1054588d4a5f/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L560
the business logic wasn't triggered but a NPE (same as https://github.com/quarkusio/quarkus/issues/24931) could have been raised
with the code in main branch

machi1990 commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:
```diff
@@ -166,6 +166,7 @@ public QuartzScheduler(SchedulerContext context, QuartzSupport quartzSupport, Sc
                         String cron = SchedulerUtils.lookUpPropertyValue(scheduled.cron());
                         if (!cron.isEmpty()) {
                             if (SchedulerUtils.isOff(cron)) {
+                                this.pause(identity);
```
Let's pause the job disabled job to effectively let Quartz know that this should not be invoked in any way. 
When the job is enabled/turned on, it'll be rescheduled and executed again.

mkouba replied:
This results in a `org.quartz.Scheduler.pauseJob(JobKey)` invocation, right? What happens if the job does not exist yet? Is it a no-op?

machi1990 replied:
If the job does not exists, it is a no-op.
I'll add a test case in the integration test project to make sure we won't have a regression here.

machi1990 commented on extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:
```diff
@@ -214,6 +215,7 @@ public QuartzScheduler(SchedulerContext context, QuartzSupport quartzSupport, Sc
                         } else if (!scheduled.every().isEmpty()) {
                             OptionalLong everyMillis = SchedulerUtils.parseEveryAsMillis(scheduled);
                             if (!everyMillis.isPresent()) {
+                                this.pause(identity);
```
Same comment as https://github.com/quarkusio/quarkus/pull/25408/files#r866356219

machi1990 commented:
@ruiarodrigues this should fix the issue reported in #25214. Can you have a look?

ruiarodrigues commented:
@machi1990 What am I suppose to do/check? 😄

machi1990 commented:
> @machi1990 What am I suppose to do/check? smile

:-) 
I've tested that the patch effectively fixes the issue. I'll like a second eye on it as well, so it is possible, can you verify that this PR fixes the issue for you?

You can clone my fork: https://github.com/machi1990/quarkus and checkout to the `fix/25214` branch.
Once that is done you can follow the instructions in https://github.com/quarkusio/quarkus/blob/main/CONTRIBUTING.md#build (except the cloning the repo part) to build a local snapshot which you can use in the reproducer.
here is how you could use the snapshot: https://github.com/quarkusio/quarkus/blob/main/CONTRIBUTING.md#usage

machi1990 commented:
> I cloned it but I'm having some problems with the download of the dependencies. Is there any snapshot available of your fork?

No, I do not have a snapshot. What dependencies issue are you encountering?

> I checked the PR and I have one question: the if(!everyMillis.isPresent()) statement already discard the job and doesn't add it to the scheduler or reschedule it. So why do you need to pause it here? Can it be read elsewhere and added to the scheduler?

This is explained in the PR description. The job is discarded and this is fine for in memory jobs but for persisted jobs quartz will try to launch them (https://github.com/quarkusio/quarkus/blob/ace0e89aca111a3eb9d2201259fd121f9d8e1cb4/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L563) - load them from the db. The way we wired the jobs in the current `main` codebase (it might be different from the version of quarkus the reproducer is running), the trigger method annotated with `@Scheduled` won't be executed - previously this would have cause a NPE (issue https://github.com/quarkusio/quarkus/issues/24931, fixed by https://github.com/quarkusio/quarkus/pull/25407) . With just that, the @Scheduled trigger will be skipped all together. 
Now the change in this PR really make sure that Quartz does not try to fire the trigger for disabled jobs. They'll be resumed automatically when the job is enabled / turned on again.

ruiarodrigues commented:
I'm not able to run the reproducer with the 999-SNAPSHOT. I think you can go ahead and I will test it again when the final release is available

machi1990 commented:
> I'm not able to run the reproducer with the 999-SNAPSHOT. I think you can go ahead and I will test it again when the final release is available

Thanks for trying anyway. I wonder what's the issue you are facing?
In either way, @mkouba would you be able to re-review and approve/merge. I think once this is in, it can still be tested using main SNAPSHOT -  https://github.com/quarkusio/quarkus/blob/main/CONTRIBUTING.md#checking-an-issue-is-fixed-in-main

mkouba commented on integration-tests/quartz/src/main/java/io/quarkus/it/quartz/DisabledMethodResource.java:
```diff
@@ -0,0 +1,28 @@
+    @Path("cron")
+    @Produces(MediaType.TEXT_PLAIN)
+    public String getCronCount() {
+        return disabledScheduledMethods.valueSetByCronScheduledMethod;
```
You shouldn't access a static variable like this...

machi1990 replied:
I didn't give this much thought since it is a variable that's there to only show that it is not initialized.

mkouba commented on integration-tests/quartz/src/main/java/io/quarkus/it/quartz/DisabledScheduledMethods.java:
```diff
@@ -0,0 +1,22 @@
+
+@ApplicationScoped
+public class DisabledScheduledMethods {
+    static String valueSetByCronScheduledMethod = "";
```
I think that we should make these variables `volatile`...

machi1990 replied:
Same reasoning as https://github.com/quarkusio/quarkus/pull/25408#discussion_r870279981

machi1990 replied:
I see your point. I've added the suggestion and added some comments to the methods

mkouba reviewed (commented):
Looks good. I've added two minor comments about the test though...

machi1990 commented:
> Looks good. I've added two minor comments about the test though...

Thanks. Can you have a look?

machi1990 commented:
Thanks @mkouba
